packages/integration_test/lib/integration_test.dart

```diff
@@ -6,70 +6,158 @@ import 'dart:async';
-  /// Sets up a listener to report that the tests are finished when everything is
-  /// torn down.
+class IntegrationTestWidgetsFlutterBinding extends LiveTestWidgetsFlutterBinding
+    implements IntegrationTestResults {
+  /// If [run] is not used, sets up a listener to report that the tests are
+  /// finished when everything is torn down.
+  ///
+  /// This functionality is deprecated – clients are expected to use [run] to
+  /// execute their tests instead.
   IntegrationTestWidgetsFlutterBinding() {
-    // TODO(jackson): Report test results as they arrive
+    if (!_isUsingLegacyReporting) {
+      // TODO(jiahaog): Point users to use the CLI https://github.com/flutter/flutter/issues/66264.
+      print('Using the legacy test result reporter, which will not catch all '
+          'errors thrown in declared tests. Consider wrapping tests with [run] '
+          'instead.');
+      return;
+    }
+
     tearDownAll(() async {
-      try {
-        // For web integration tests we are not using the
-        // `plugins.flutter.io/integration_test`. Mark the tests as complete
-        // before invoking the channel.
-        if (kIsWeb) {
-          if (!_allTestsPassed.isCompleted) {
-            _allTestsPassed.complete(true);
-          }
-        }
-        callbackManager.cleanup();
-        await _channel.invokeMethod<void>(
-          'allTestsFinished',
-          <String, dynamic>{
-            'results': results.map((String name, Object result) {
-              if (result is Failure) {
-                return MapEntry<String, dynamic>(name, result.details);
-              }
-              return MapEntry<String, Object>(name, result);
-            })
-          },
-        );
-      } on MissingPluginException {
-        print('Warning: integration_test test plugin was not detected.');
-      }
-      if (!_allTestsPassed.isCompleted) {
-        _allTestsPassed.complete(true);
-      }
+      _updateTestResultState(results);
+      await const _ReporterImpl().report(results);
     });
 
-    // TODO(jackson): Report the results individually instead of all at once
-    // See https://github.com/flutter/flutter/issues/38985
     final TestExceptionReporter oldTestExceptionReporter = reportTestException;
     reportTestException =
         (FlutterErrorDetails details, String testDescription) {
-      results[testDescription] = Failure(testDescription, details.toString());
-      if (!_allTestsPassed.isCompleted) {
-        _allTestsPassed.complete(false);
-      }
+      results[testDescription] = Failure(testDescription, details.toString(),
+          error: details.exception);
       oldTestExceptionReporter(details, testDescription);
     };
   }
 
+  void _updateTestResultState(Map<String, Object> results) {
+    this.results = results;
+    print('Test execution completed: $results');
+
+    _allTestsPassed
+        .complete(!results.values.any((Object val) => val is Failure));
```

Comment:
nit: formatting

Is there any way we could avoid switching on types here? That can become brittle. Could we instead have the `results` be of such a type that it has some property saying whether it's an error?

Maybe rather than a separate `Failure` type, we have a `.failure` ctor?

Reply:
Assuming you mean something like this: 

```dart
enum ResultStatus {
  pass, fail
}

class Result {
  final ResultStatus status;

  Result.failure(...);

  Result.success(...);
}
```
I'm all for this, but the inconvenient bit is that `Failure` today is already part of the public type, and we might not want to replace Failure with the Result above.

Dart doesn't have typed unions, but another option would be to have some sort of superclass type:

```dart
class Result {}

// make the old failure extend [Result]
class Failure extends Result {}

class Success extends Result {}
```
And then store them in the binding as `Map<String, Result>`, or even `List<Result>` (since the key for the test name can be contained within the class). What do you think?

Reply:
I think having the supertype is fine. We could then just add some bool getter on it about whether the result describes a failure, and make `Failure` implement it as true.

Reply:
Added the supertype. I think the bool getter is a bit unnecessary and switching on types should be alright – now that the parent type is `Result` checking for `Failure` should be safer